### PR TITLE
HPCC-30349 Add open telemetry support to esp

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/services/ws_ecl/ws_ecl_service.cpp [hunk 1]
```diff
@@ -1993,20 +1993,18 @@ int CWsEclBinding::submitWsEclWorkunit(IEspContext & context, WsEclWuInfo &wsinf
...
+    ISpan * activeSpan = context.queryActiveSpan();
+    if (activeSpan)
     {
-        StringBuffer globalId, callerId;
-        wsecl->getHttpGlobalIdHeader(httpreq, globalId);
-        wsecl->getHttpCallerIdHeader(httpreq, callerId);
-        if (globalId.length())
-        {
-            workunit->setDebugValue("GlobalId", globalId.str(), true);
+        clientSpan.setown(activeSpan->createClientSpan("wsecl"));
```
Comment (ghalliday) [inline esp/services/ws_ecl/ws_ecl_service.cpp:2001]: @afishbeck I think this code was wrong before - I didn't pick up the local id (or global id?) assigned if none was provided by the http request.

Comment (afishbeck) [reply esp/services/ws_ecl/ws_ecl_service.cpp:2001]: This probably wasn't tested much when it came to global-ids but is a good fix, and we definitely want it working for trace/span support.

Comment (afishbeck) [reply esp/services/ws_ecl/ws_ecl_service.cpp:2001]: @ghalliday but a more descriptive name for the span might be good.  Something like "wsecl/workunit" or similar.  We should probably try and be consistent in general format.  I stole that particular format from the opentelemetry demo which has one of it's services written in c++.  https://github.com/open-telemetry/opentelemetry-demo/blob/fa529260273fa53b4e018daa73f30b1d30b88b3a/src/currencyservice/src/server.cpp#L112

That service seems to use the form `"<servicename>/<operation>"`.  For our client spans I think something like may work.

File: system/jlib/jprop.cpp [hunk 2]
```diff
@@ -402,6 +402,40 @@ IProperties *cloneProperties(const IProperties * source, bool nocase)
...
+        if ((delineator == nullptr) || (delineator == httpHeader))
+            continue;
+
+        const char * value = delineator + 1;
+        while (isspace(*value))
+            value++;
+
+        if (*value)
+        {
+            key.clear().append(delineator - httpHeader, httpHeader);
```
Comment (afishbeck) [inline system/jlib/jprop.cpp:425]: Not necessary for our use cases, but would be pretty easy to remove trailing whitespace before dlineator if you wanted to be flexible about supporting "x = y".  Or would this be a case of being strict is actually better?

**Review by afishbeck — COMMENTED** [review]
@ghalliday looks good, some comments.

**Comment by afishbeck** [comment]
@ghalliday I think as a follow on to instrumenting roxie/thor and esp we should consider some quick win attributes to the spans that may be very helpful.  We'll probably enhance them greatly over time.. but some quick wins like general status of the span for example.  Any exception that caused the span to end would be very useful.

File: system/jlib/jprop.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jlib/jprop.cpp:425]: I think revisit if there is a use-case it would actually be useful for.

**Review by rpastrana — COMMENTED** [review]
@ghalliday a couple of comments

File: esp/bindings/http/platform/httptransport.cpp [hunk 3]
```diff
@@ -1940,17 +1940,11 @@ void CHttpRequest::updateContext()
...
-            m_context->setGlobalId(globalId);
-        getHeader(HTTP_HEADER_HPCC_CALLER_ID, callerId);
-        if (callerId.isEmpty())
-            getHeader("hpcc-caller-id", callerId);
-        if(callerId.length())
-            m_context->setCallerId(callerId);
+
+        //MORE: The previous code would be better off querying httpHeaders...
+        Owned<IProperties> httpHeaders = getHeadersAsProperties(m_headers);
+        Owned<ISpan> requestSpan = queryTraceManager().createServerSpan("request", httpHeaders, SpanFlags::EnsureGlobalId);
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply esp/bindings/http/platform/httptransport.cpp:1946]: I think it is likely the rest of the code would be better using an IProperties - so I will leave.

File: esp/services/esdl_svc_engine/esdl_binding.cpp [hunk 4]
```diff
@@ -1571,9 +1571,12 @@ void EsdlServiceImpl::sendTargetSOAP(IEspContext & context,
...
     }
 
-    Owned<IProperties> headers = createProperties();
-    headers->setProp(HTTP_HEADER_HPCC_GLOBAL_ID, context.getGlobalId());
-    headers->setProp(HTTP_HEADER_HPCC_CALLER_ID, context.getLocalId());
+    Owned<ISpan> clientSpan;
+    ISpan * activeSpan = context.queryActiveSpan();
+    if (activeSpan)
+        clientSpan.setown(activeSpan->createClientSpan("soapcall"));
+
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply esp/services/esdl_svc_engine/esdl_binding.cpp:1578]: There may be no span if the request comes through on a protocol that doesn't support it.  We should be creating server spans in those cases as well, but until we can guarantee they are all covered it is safer to check 1st.

File: system/jlib/jtrace.cpp [hunk 5]
```diff
@@ -235,19 +235,19 @@ class CSpan : public CInterfaceOf<ISpan>
         if (ctxProps == nullptr)
             return false;
 
-        if (!isEmptyString(hpccGlobalId.get()))
-            ctxProps->setProp(kGlobalIdHttpHeaderName, hpccGlobalId.get());
+        const char * globalId = queryGlobalId();
+        ctxProps->setNonEmptyProp(kGlobalIdHttpHeaderName, globalId);
```
Comment (afishbeck) [inline system/jlib/jtrace.cpp:239]: Trivial, but I think you could now get rid of the variable globalId and write:
`ctxProps->setNonEmptyProp(kGlobalIdHttpHeaderName, queryGlobalId());`

**Review by afishbeck — APPROVED** [review]
Approved, but one trivial comment that could get rid of a variable.

**Pull request merged**